### DISCUSSION TITLE
GrainSerializer base types

### DISCUSSION THREAD

**Discussion by jbockle** [dotnet/orleans#7597] [Q&A] [answered]
when using `GrainSerializer` with inherited types, should the `IdAttribute` argument increment after the base type?

e.g. 
```cs
[GenerateSerializer]
public abstract class Person
{
    [Id(0)]
    public string FirstName { get; set; }
}

[GenerateSerializer]
public class Employee : Person
{
    [Id(0)] // ---> should this be 1?
    public string Department { get; set; }  
}
```

**Comment by ReubenBond** [comment, marked as answer]
They do *not* share id ranges, so it's entirely fine to start from zero in each individual type (in fact, it's preferable because smaller numbers take less space).

If you have a type like A : B : C, the way it's encoded on the wire is roughly this:
`<Start> ... C fields ... <EndBase> ... B fields .... <EndBase> ... A fields ... <End>`

**Reply by jbockle** [reply]
good to know - thanks!
